packages/autorest.python/autorest/preprocess/__init__.py

```diff
@@ -307,19 +320,36 @@ def update_client(self, yaml_data: Dict[str, Any]) -> None:
         yaml_data["builderPadName"] = to_snake_case(prop_name)
         for og in yaml_data["operationGroups"]:
             for o in og["operations"]:
-                for p in o["parameters"]:
+                idx_if_match = -1
+                idx_if_none_match = -1
+                for idx, p in enumerate(o["parameters"]):
+                    wire_name_lower = (p.get("wireName") or "").lower()
                     if (
                         p["location"] == "header"
-                        and p["wireName"] == "client-request-id"
+                        and wire_name_lower == "client-request-id"
                     ):
                         yaml_data["requestIdHeaderName"] = p["wireName"]
-                    if (
-                        self.version_tolerant
-                        and p["location"] == "header"
-                        and p["clientName"] in ("if_match", "if_none_match")
-                    ):
-                        o["hasEtag"] = True
-                        yaml_data["hasEtag"] = True
+                    if self.version_tolerant and p["location"] == "header":
+                        if wire_name_lower == "if-match":
+                            idx_if_match = idx
+                        if wire_name_lower == "if-none-match":
+                            idx_if_none_match = idx
+                if idx_if_match > 0 or idx_if_none_match > 0:
+                    o["hasEtag"] = True
+                    yaml_data["hasEtag"] = True
+                    # pylint: disable=line-too-long
+                    # some service(e.g. https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cosmos-db/data-plane/Microsoft.Tables/preview/2019-02-02/table.json)
+                    # only has one, so we need to add "if-none-match" or "if-match" if it's missing
+                    add_missing_parameter(
+                        o["parameters"], idx_if_match, idx_if_none_match, "if-match"
+                    )
+                    add_missing_parameter(
```

Comment:
Will the idx change after first insert?

Reply:
We do insert after whole loop and there will be only 0 or 1 time to insert.

Reply:
> We do insert after whole loop and there will be only 0 or 1 time to insert.

 do you mean the two add_xxx will not do insert at the same time?

Reply:
> > We do insert after whole loop and there will be only 0 or 1 time to insert.
> 
> do you mean the two add_xxx will not do insert at the same time?

Yep